gitlab-runner/src/lib.rs

```diff
@@ -105,34 +106,122 @@ macro_rules! outputln {
 pub type JobResult = Result<(), ()>;
 pub use client::Phase;
 
+/// Async trait for handling a single Job that handles its own cancellation
+///
+/// This trait is largely identical to [`JobHandler`], but its methods explicitly take a
+/// `CancellationToken`, which will be triggered if GitLab cancels the job, after which the method
+/// will be responsible for cancellation appropriately. In most cases, the entire execution should
+/// simply be cancelled, in which case [`JobHandler`]'s default behavior is desirable instead. (Even
+/// when cancelled, `cleanup` will still be called, allowing any cleanup tasks to be performed.)
+///
+/// Note that this is an asynchronous trait which should be implemented by using the [`async_trait`]
+/// crate. However this also means the rustdoc documentation is interesting...
+#[async_trait::async_trait]
+pub trait CancellableJobHandler: Send {
+    /// Do a single step of a job
+    ///
+    /// This gets called for each phase of the job (e.g. script and after_script). The passed
+    /// string array is the same array as was passed for a given step in the job definition. If the
+    /// job is cancelled while this is running, the given `cancel_token` will be triggered.
+    ///
+    /// Note that gitlab concatinates the `before_script` and `script` arrays into a single
+    /// [Phase::Script] step
+    async fn step(
+        &mut self,
+        script: &[String],
+        phase: Phase,
+        cancel_token: &CancellationToken,
+    ) -> JobResult;
+    /// Upload artifacts to gitlab
+    ///
+    /// This gets called depending on whether the job definition calls for artifacts to be uploaded
+    /// based on the result of the script run. If the job is cancelled while this is running, the
+    /// given `cancel_token` will be triggered.If the job was already cancelled, this will not be
+    /// called.
+    async fn upload_artifacts(
+        &mut self,
+        _uploader: &mut Uploader,
+        _cancel_token: &CancellationToken,
```

Comment:
Does it maek sense to add the cancellation token in this one? This should only be called when the job is done and the final artifacts are (being) uploaded. Seems little point in that still being a cancellation point.

Reply:
It would be nice if e.g. uploading relatively large artifacts, but I'm also not really tied to it and can remove it if still desired.

Reply:
I think it's worth finishing the upload once it's started rather then potentially cutting things of at half the upload. It's then up to gitlab what to do with the artifacts really.